services/wallet/api.go

```diff
@@ -293,3 +304,74 @@ func (api *API) GetSuggestedFees(ctx context.Context, chainID uint64) (*Suggeste
 	log.Debug("call to GetSuggestedFees")
 	return api.s.feesManager.suggestedFees(ctx, chainID)
 }
+
+func (api *API) GetDerivedAddressesForPath(password string, derivedFrom string, path string, pageSize int, pageNumber int) ([]*DerivedAddress, error) {
+	info, err := api.s.gethManager.AccountsGenerator().LoadAccount(derivedFrom, password)
+	if err != nil {
+		return nil, err
+	}
+
+	return api.getDerivedAddresses(info.ID, path, pageSize, pageNumber)
+}
+
+func (api *API) GetDerivedAddressesForMenominicWithPath(mnemonic string, path string, pageSize int, pageNumber int) ([]*DerivedAddress, error) {
+	mnemonicNoExtraSpaces := strings.Join(strings.Fields(mnemonic), " ")
+
+	info, err := api.s.gethManager.AccountsGenerator().ImportMnemonic(mnemonicNoExtraSpaces, "")
+	if err != nil {
+		return nil, err
+	}
+
+	return api.getDerivedAddresses(info.ID, path, pageSize, pageNumber)
+}
+
+func (api *API) getDerivedAddresses(id string, path string, pageSize int, pageNumber int) ([]*DerivedAddress, error) {
+	addedAccounts, err := api.s.db.GetAccounts()
+	if err != nil {
+		return nil, err
+	}
+
+	derivedAddresses := make([]*DerivedAddress, 0)
+
+	if pageNumber <= 0 || pageSize <= 0 {
+		return nil, fmt.Errorf("pageSize and pageNumber should be greater than 0")
+	}
+
+	var startIndex = ((pageNumber - 1) * pageSize)
+	var endIndex = (pageNumber * pageSize)
+
+	for i := startIndex; i < endIndex; i++ {
+		derivedPath := fmt.Sprint(path, "/", i)
+
+		info, err := api.s.gethManager.AccountsGenerator().DeriveAddresses(id, []string{derivedPath})
+		if err != nil {
+			return nil, err
+		}
+
+		alreadyExists := false
+		for _, account := range addedAccounts {
+			if types.Address(common.HexToAddress(info[derivedPath].Address)) == account.Address {
+				alreadyExists = true
+				break
+			}
+		}
+
+		var ctx context.Context
+		transactions, err := api.s.transferController.GetTransfersByAddress(ctx, api.s.rpcClient.UpstreamChainID, common.HexToAddress(info[derivedPath].Address), nil, (*hexutil.Big)(big.NewInt(1)), false)
```

Comment:
I think this loop and in particular this call should be made in a coroutine to execute all the calls in parallel. If that helps, you can see how it is done here: https://github.com/status-im/status-go/blob/develop/services/wallet/token.go#L245

Reply:
hey thanks for pointing to an example. It was just really quick when I tested it and didnt do it. Also do you mean https://github.com/status-im/status-go/blob/d884749ebf5e4edff46295c27e583425820e64ac/services/wallet/token.go#L260? cause this qpi getTokenBalance is just a normal function right?

Reply:
hey made getting each of the derived address concurrent ! @alaibe